### ISSUE TITLE
Injectable provideIn useFactory, stop resolving constructor parameters

### ISSUE THREAD

**Issue by kukjevov** [angular/angular#24126]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior

When using `Injectable` with `provideIn` and `useFactory` `ngc` shows following warning
> Warning: Can't resolve all parameters for AuthenticationService in D:/git/ng/ng-authorization/src/common/authentication.service.ts: (?). This will become an error in Angular v6.x

When this was done without so called *tree-shakeable* provider it was done by defining `FactoryProvider` and `AuthenticationService ` service itself was not *flagged* with `Injeactable` because it was created only in factory function, that means parameters for constructors were provided by factory function which had its own *resolvable* parameters. So this error could not happen.

If i want to use new *tree-shakeable* provider for this `AuthenticationService` i had to do some *hacking*, because type of parameter of constructor for this class is *interface* so it can not be resolved. As soon as i put `Injectable` decorator on this class i get earlier mentioned warning which is not nice, since this class is never created by `Injector` but actualy by factory function. 

This should have been checked if there is `useFactory`, if so do not check constructor parameters :).

## Minimal reproduction of the problem with instructions
**Original Solution** *working one*

```ts
export interface AuthenticationServiceOptions
{
}

export class AuthenticationService
{
    constructor(private _options: AuthenticationServiceOptions)
    {
    }
}

export function authenticationServiceFactory()
{
    return new AuthenticationService({});
}

export const AUTHENTICATION_SERVICE_PROVIDER: FactoryProvider =
{ 
    provide: AuthenticationService,
    useFactory: authenticationServiceFactory
};
```

**New Solution** *working one with hack*

```ts
export interface AuthenticationServiceOptions
{
}

export function authenticationServiceFactory()
{
    return new AuthenticationService({});
}

@Injectable({providedIn: 'root', useFactory: authenticationServiceFactory})
export class AuthenticationService
{
    constructor(@Inject(Injector) private _options: AuthenticationServiceOptions)
    {
    }
}
```

## What is the motivation / use case for changing the behavior?

I would like to use new *tree-shakeable* providers with `useFactory` without hacking of constructor parameters.

## Environment

<pre><code>
Angular version: 6.0.3

Browser: ALL
- [ ] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: 9.4.0
- Platform:  Windows

Others:

</code></pre>

**Comment by Johanneke** [comment]
I feel like this will solve my issue as well, where an `Injectable` has a string as parameter, is provided with use of a factory method, but still gives a warning ```Warning: Can't resolve all parameters for MyService in .... This will become an error in Angular v6.x```
I don't see why this warning would appear at all, and I really don't want an error in the future. Am I right in thinking that this reported issue is the same as what I describe?

**Comment by alxhub** [comment]
Closing as this is fixed in 9.0 (Ivy)! `@Injectable({useFactory: ...})` doesn't require constructor parameters to be statically resolvable.

**Comment by kukjevov** [comment]
@alxhub hmm are you sure that this is fixed ?

I have following class
``` ts
/**
 * Service used as configuration for *console sink component*
 */
@Injectable({providedIn: 'root', useFactory: () => new ConsoleSinkConfigService()})
export class ConsoleSinkConfigService
{
    //######################### constructor #########################

    /**
     * Creates instance of `ConsoleSinkConfigService`
     * @param maxLogsCount - Maximal number of logs that can be stored
     * @param restrictToLevel - Minimal log level to be displayed
     */
    constructor(public maxLogsCount?: number, public restrictToLevel?: LogEventLevel)
    {
        if(isBlank(maxLogsCount))
        {
            this.maxLogsCount = MAX_LOGS;
        }

        if(isBlank(restrictToLevel))
        {
            this.restrictToLevel = LogEventLevel.verbose;
        }
    }
}
```

and i got following *warning* consoleSinkConfig.service.ts: (?, ?). This will become an error in Angular v6.x

But i have `"enableIvy": false` since it is library, as far as i know ivy should not be enable for library.
